src/fsharp/FSharp.Core/seqcore.fs

```diff
@@ -261,7 +260,7 @@ namespace Microsoft.FSharp.Core.CompilerServices
                 member x.Current = box (x.GetCurrent())
 
                 member x.MoveNext() =
-                   if not started then (started <- true)
+                   if not started then started <- true
                    if finished then false
```

Comment:
I'm wondering if we could swap finished check first, I assume finished being true means started is true, so it could be the first thing checked and avoid the `not started set started` check.

I guess the impact of such change would warrant further impact study and a separate PR, but maybe somebody knows if such change would be wrong or if it sounds interesting to look further?

Reply:
@smoothdeveloper 

From:  https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerator?view=netframework-4.7.2

````
Initially, the enumerator is positioned before the first element in the collection. You must call the MoveNext method to advance the enumerator to the first element of the collection before reading the value of Current; otherwise, Current is undefined. 
````
I believe this ensures that.